### PR TITLE
Scripts for verifying bytecodes

### LINKED ISSUE #5082: Provide script to verify git branch matches deployed contracts system
### Expected Behavior

`yarn verify-bytecodes -b BRANCH -n NETWORK` verifies that the Solidity source codes under `packages/protocol/contracts`.on branch BRANCH match the bytecodes deployed to NETWORK.

### Current Behavior

No such script

### LINKED ISSUE #5083: Provide script to verify git branch matches proposed contracts system
### Expected Behavior

`yarn verify-proposal -b BRANCH -n NETWORK -p PROPOSAL` verifies that the Solidity source codes under `packages/protocol/contracts.on` branch BRANCH match the bytecodes deployed to NETWORK, including any changes that PROPOSAL would cause.

### Current Behavior

No such script.

### PR DESCRIPTION
**Pull request by m-chrzan**
### Description

Adds two bash wrappers around `verify-bytecodes`:

* `verify-deployed` checks that the current Celo contracts correspond to Solidity sources on a given branch.
* `verify-release` checks that after executing a given proposal, the onchain contracts will correspond to Solidity sources on a given branch.

### Other changes

Added a `-f` flag to several of our bash scripts, and modified `truffle-config` to take a `--forno` parameter that allows Truffle scripts to run with a Forno provider on testnets/the mainnet for which a Forno instance exists.

Added comments, cleaned up scripts.

#### verify-deployed
    celotooljs port-forward -e baklava
    yarn verify-deployed -n baklava -b baklava -f

#### verify-release

    git checkout rc1
    yarn devchain generate-tar devchain.tar.gz
    git checkout -
    yarn devchain run-tar devchain.tar.gz
    yarn make-release -a rc1 -b m-chrzan/bytecode-scripts -n development -p proposal.json -i example-initialize-data.json -r report.json
    yarn verify-release -b m-chrzan/bytecode-scripts -n development -p proposal.json

### Related issues

- Fixes #5082
- Fixes #5083

### Backwards compatibility

+1

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -491,6 +491,56 @@ jobs:
...
+      - run:
+          name: Verify bytecodes between RC1 and current branch
+          command: |
+            cd packages/protocol
+
+            # deploy RC1 contracts to ganache devchain
+            # We fetch via HTTPS instead of SSH to avoid the interactive "do you
+            # trust the RSA key fingerprint" prompt.
+            git remote add origin-https https://github.com/celo-org/celo-monorepo.git
+            git fetch origin-https rc1
```
Comment (nambrot) [inline .circleci/config.yml:516]: At this point, this is incorrect, as `master` needs to compare against release 1.

I believe there are only 3 options:

1. Adjust the migrations to proxy linked libraries and then checkout the release-1 branch here
2. Create a special case for release 2, where creating the bytecode involves 
  2.1 Deploying RC1
  2.2 Creating the proposal for release 1
  2.3 Get through the proposal for release 1
  2.4 Continue
3. Not add this to CI until 1. is done

File: .circleci/config.yml [hunk 2]
```diff
@@ -491,6 +491,56 @@ jobs:
...
+    resource_class: large
+    steps:
+      - attach_workspace:
+          at: ~/app
+      - run:
+          name: Check if the test should run
+          command: |
+            ./scripts/ci_check_if_test_should_run_v2.sh @celo/protocol
+      - run:
+          name: Verify bytecodes between RC1 and current branch
```
Comment (nambrot) [inline .circleci/config.yml]:
```suggestion
          name: Verify bytecodes between most recent release and current branch
```

File: packages/docs/celo-codebase/protocol/governance.md [hunk 3]
```diff
@@ -36,11 +36,27 @@ In order for a proposal to pass, it must meet a minimum threshold for **particip
 * Participation is the minimum portion of Locked Gold which must cast a vote for a proposal to pass. It exists to prevent proposals passing with very low participation. The participation requirement is calculated as a governable portion of the participation baseline, which is an exponential moving average of final participation in past governance proposals.
 * Agreement is the portion of votes cast that must be "yes" votes for a proposal to pass. Each contract and function can define a required level of agreement, and the required agreement for a proposal is the maximum requirement among its constituent transactions.
 
-## Execution
+### Execution
 
 Proposals that graduate from the Referendum phase to the execution phase may be executed by anyone, triggering a “call” operation code with the arguments defined in the proposal, originating from the `Governance` smart contract. Proposals expire from this phase after two days.
 
-## Hotfix
+## Smart Contract Upgradeability
```
Comment (nambrot) [inline packages/docs/celo-codebase/protocol/governance.md:43]: 🙏

File: packages/docs/community/release-process/smart-contracts.md [hunk 4]
```diff
@@ -46,57 +46,84 @@ Github branches/tags and Github releases are used to coordinate past and ongoing
 2. On master branch, `.circleci/config.yml` should be edited so that the variable `RELEASE_TAG` points to the tag `celo-core-contracts-v${N}.mainnet`
 
 
-## Build and Release Process
+## Release Process
 
-Using the following scripts, new contracts can be built and deployed alongside a corresponding on-chain governance proposal. Run the following scripts from the branch with the new release contracts:
+The cLabs team has provided several scripts (under `packages/protocol` in [celo-org/celo-monorepo](https://github.com/celo-org/celo-monorepo) and via [celocli](../../command-line-interface/introduction.md)) for use in the release process and with contract upgrade governance proposals to give participating stakeholders increased confidence.
```
Comment (nambrot) [inline packages/docs/community/release-process/smart-contracts.md]: Unnecessarily mentioning cLabs here IMO

File: packages/docs/community/release-process/smart-contracts.md [hunk 5]
```diff
@@ -46,57 +46,84 @@ Github branches/tags and Github releases are used to coordinate past and ongoing
...
-## Script 2 ##
-# TODO: command to pull source code from current release branch
-# generate report on backwards compatibility between RC1 and current branch with new release contracts
-yarn ts-node scripts/check-backward.ts report --exclude ".*Test|Mock.*|I[A-Z].*|.*Proxy|LinkedList|SortedLinkedList|SortedLinkedListWithMedian|MultiSig.*|ReleaseGold" -o build/rc1 -n build/contracts --output_file report.json
-
-## Script 3 ##
-# deploy new contracts and generate governance proposal for upgrade
-yarn truffle exec --network development ./scripts/truffle/make-release.js --build_directory build/ --report report.json --proposal proposal.json --initialize_data example-initialize-data.json'
+NETWORK=${"baklava"|"alfajores"|"mainnet"}
+RELEASE="celo-core-contracts-v${N}.${NETWORK}"
```
Comment (nambrot) [inline packages/docs/community/release-process/smart-contracts.md]:
```suggestion
RELEASE="celo-core-contracts-v${N-1}.${NETWORK}"
```

I think?

File: packages/docs/community/release-process/smart-contracts.md [hunk 6]
```diff
@@ -46,57 +46,84 @@ Github branches/tags and Github releases are used to coordinate past and ongoing
...
+# A -f boolean flag can be provided to use a forno full node to connect to the provided network
+# A -r boolean flag should be provided if this is the first release (before linked libraries were proxied)
+yarn verify-deployed -n $NETWORK -b $RELEASE -r -f
```
Comment (nambrot) [inline packages/docs/community/release-process/smart-contracts.md:67]: I don't think it's in scope for this PR, but for these critical scripts, I'm not sure I'm a fan of these shortened flags, wdyt?

File: packages/docs/community/release-process/smart-contracts.md [hunk 7]
```diff
@@ -46,57 +46,84 @@ Github branches/tags and Github releases are used to coordinate past and ongoing
...
 
-1. Compiles the contracts at <code>rc1</code> branch and confirms that the compiled bytecode matches what is currently deployed on the specified network.
+```bash
+NETWORK=${"baklava"|"alfajores"|"mainnet"}
+PREVIOUS_RELEASE="celo-core-contracts-v${N-1}.${NETWORK}"
+RELEASE_CANDIDATE="celo-core-contracts-v${N}.rc${X}"
+yarn check-versions -a $PREVIOUS_RELEASE -b $RELEASE_CANDIDATE -r "report.json"
+```
+
+This should be used in tandem with `verify-deployed -b $PREVIOUS_RELEASE -n $NETWORK` to ensure the compatibility checks compare the release candidate to what is actually active on the network.
```
Comment (nambrot) [inline packages/docs/community/release-process/smart-contracts.md:85]: 🙏

File: packages/protocol/scripts/bash/check-versions.sh [hunk 8]
```diff
@@ -7,19 +7,16 @@ set -euo pipefail
 # Flags:
 # -a: Old branch containing smart contracts, which has likely been released.
 # -b: New branch containing smart contracts, on which version numbers may be updated.
-# -n: The network to deploy to.
 # -r: Path that the contract compatibility report should be written to.
 
 BRANCH_1=""
```
Comment (nambrot) [inline packages/protocol/scripts/bash/check-versions.sh]: Would be nice to rename Branch 1/2 to something more semantic

File: packages/protocol/scripts/truffle/verify-bytecode.ts [hunk 9]
```diff
@@ -5,6 +5,30 @@ import { ProxyInstance, RegistryInstance } from 'types'
 
 import fs = require('fs')
 
+/*
```
Comment (nambrot) [inline packages/protocol/scripts/truffle/verify-bytecode.ts:8]: 🙏

**Review by nambrot — COMMENTED** [review]
Amazing!

File: packages/docs/community/release-process/smart-contracts.md (hunk 6 above)
Comment (yorhodes) [reply packages/docs/community/release-process/smart-contracts.md:67]: agreed

File: packages/docs/community/release-process/smart-contracts.md (hunk 5 above)
Comment (yorhodes) [reply packages/docs/community/release-process/smart-contracts.md]: I was trying to generalize this and add the specific instruction in backward compatibility check but this is probably the primary use case

File: packages/docs/community/release-process/smart-contracts.md (hunk 6 above)
Comment (yorhodes) [reply packages/docs/community/release-process/smart-contracts.md:67]: will address in followup PR

**Review by yorhodes — APPROVED** [review]
LGTM

File: .circleci/config.yml (hunk 1 above)
Comment (yorhodes) [reply .circleci/config.yml:516]: went with option 3

**Review by nambrot — APPROVED** [review]

**Pull request merged**